reasoner/planner/ReasonerPlanner.java

```diff
@@ -0,0 +1,144 @@
+                    resolvables.addAll(conjunction.negations());
+                    compiled.put(conjunction, resolvables);
+                }
+            }
+        }
+        return compiled.get(conjunction);
+    }
+
+    public Plan<Resolvable<?>> getPlan(ResolvableConjunction conjunction, Set<Identifier.Variable.Retrievable> bounds) {
+        Pair<ResolvableConjunction, Set<Identifier.Variable.Retrievable>> plannableKey = new Pair<>(conjunction, bounds);
+        if (!planCache.containsKey(plannableKey)) {
+            synchronized (planCache) {
+                if (!planCache.containsKey(plannableKey)) {
+                    planCache.put(plannableKey, planConjunction(conjunction, bounds));
+                }
+            }
+        }
+        return planCache.get(plannableKey);
+    }
+
+    protected abstract Plan<Resolvable<?>> planConjunction(ResolvableConjunction conjunction, Set<Identifier.Variable.Retrievable> bounds);
+
+    protected boolean dependenciesSatisfied(Resolvable<?> resolvable, Set<Identifier.Variable.Retrievable> bounds, Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies) {
+        return bounds.containsAll(dependencies.get(resolvable));
+    }
+
+    /**
+     * Determine the resolvables that are dependent upon the generation of each variable
+     */
+    protected static Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> dependencies(Set<Resolvable<?>> resolvables) {
+        Map<Resolvable<?>, Set<Identifier.Variable.Retrievable>> deps = new HashMap<>();
+        Set<Identifier.Variable.Retrievable> generated = iterate(resolvables).map(Resolvable::generating).filter(Optional::isPresent)
+                .map(Optional::get).map(ThingVariable::id).toSet();
+
+        Map<Identifier.Variable.Retrievable, Integer> unnegatedRefCount = new HashMap<>();
+        for (Resolvable<?> resolvable : resolvables) {
+            Optional<ThingVariable> generating = resolvable.generating();
+            deps.putIfAbsent(resolvable, new HashSet<>());
+            for (Identifier.Variable.Retrievable v : resolvable.retrieves()) {
+                if (generated.contains(v) && !(generating.isPresent() && generating.get().id().equals(v))) {
+                    // TODO: Should this rule the Resolvable<?> out if generates it's own dependency?
+                    deps.get(resolvable).add(v);
+                }
+                if (!resolvable.isNegated()) {
+                    unnegatedRefCount.put(v, 1 + unnegatedRefCount.getOrDefault(v, 0));
+                }
+            }
+        }
+
+        for (Resolvable<?> resolvable : resolvables) {
+            if (resolvable.isNegated()) {
+                for (Identifier.Variable.Retrievable v : resolvable.retrieves()) {
+                    if (unnegatedRefCount.getOrDefault(v, 0) > 0) {
+                        deps.get(resolvable).add(v);
+                    }
+                }
+            }
+        }
```

Comment:
If a variable in a negated appears unnegated in some concludable, the negated is dependent on that concludable.
Here, circular dependencies are impossible since negated cannot be 'generating' any variables